### PR TITLE
Factorissimo Improvements

### PR DESCRIPTION
**Pull request by Olipro**
Hi there,

I originally set out to add the ability to use inserters with chests and then after looking over the code, I felt it could do with a bit of reorganisation and I now have it at a point where it's at the very least worth taking a look at.

Since I haven't yet implemented migrations or the remote interface, it's definitely not ready for primetime, but there should be enough here already to make it worthwhile to build upon. Some of the plumbing code in lib/ is a *bit* hard to understand (I am of course happy to explain it), but the functionality it provides makes building upon the object types very trivial and I don't personally expect there to be too much necessary in terms of modifying it.

### CODE REVIEW

**Comment by Tomirad** [comment]
What with fork of large structures from factories ReverseStateMonad?
https://github.com/ReverseStateMonad/Factorissimo/tree/bfde78ca7518b989c9ca2aab9460d36c9456a30c

**Comment by Olipro** [comment]
Firstly, does your question have anything to do with this pull request?

Secondly, please could you rephrase your question in coherent English so that I can understand it.

**Comment by Tomirad** [comment]
I present a proposal for the next improvements Factorissimo.

**Comment by ReverseStateMonad** [comment]
I've been looking at what it would take to make the two compatible; it should be possible, although it might take a while.  While I was at it, I found some code in table-utils that could stand to be improved (one cleanup & one bugfix).  I'll have it up in a bit.

**Comment by Olipro** [comment]
You should be able to add larger buildings purely by defining a new layout - there may be a few locations where I omitted to generify the code in which case that should be fixed up, but generally speaking, it should be independent of the implementation class unless you happen to be introducing type-specific features to it.

Also, I took a look through your commits and they're quite numerous - before pushing, consider using `git rebase -i` to squash your commits when they're a sequence of minor changes made chronologically close together.

**Comment by ReverseStateMonad** [comment]
That was just me fighting with Github's web interface... it literally wouldn't let me add everything at once, so I had to split up a few large commits into chunks it would accept.  When I'm not using the web interface it's not an issue.
My branch isn't just adding bigger buildings, though - I also redid the buildings to be able to face in different directions and so that pipe connections would work better, and rewrote a bunch of the layout code to be able to automate building layout generation (one generating function instead of a hand-written layout for each building).  That's where all of the changes to make the branches compatible come in; references need to be updated since the layout of the data structure is a little different, and there are a number of assumptions your code (and the original factorissimo code) makes that are specific to the direction the building is facing and a few that are specific to the size.

PS - take a look at https://github.com/ReverseStateMonad/Factorissimo/commit/b9638552d238b46eeb031638cd29e7053c064ff3 it's not specific to my building changes, just improvements to your table utils.

Edit - looking at it again, I misread the purpose of remove_by_pred; I was expecting it to remove ALL matching elements, not just the first one.

**Comment by Olipro** [comment]
Yes, I looked through the commit and left some comments.

Table changes look fine and certainly necessary, You can send a pull request or I can just cherry-pick the commit straight onto my branch if you prefer.

**Comment by MagmaMcFry** [comment]
Wow, this is a pretty major rewrite. Olipro, you've certainly cleaned the code up. In fact, you've cleaned it up so much that I have barely any idea what happens where anymore, and I wouldn't be able to work with your code even if I had any intent to do so. I know my own code is messy, since it is still almost the same old proof-of-concept code that I wrote in three nights without any future-proofing in mind, but unless your rewrite magically increases Factorissimo's performance by a factor of at least four, I do not wish to merge this, since it breaks save compatibility as well as my own compatibility in favor of structurally sound yet (imo) ridiculously overengineered code.

So kudos on your software engineering skills, but I'm afraid I have to decline on this merge. My plan is and has been for a while to do a complete rewrite of Factorissimo myself (in order to fix bugs and add features that were simply not feasible with the current code), and I hope you'll be more happy with the more well-designed Lua code that will come with my rewrite.

**Comment by Olipro** [comment]
Well, as I originally stated when submitting the Pull Request:

> Some of the plumbing code in lib/ is a bit hard to understand (I am of course happy to explain it), but the functionality it provides makes building upon the object types very trivial and I don't personally expect there to be too much necessary in terms of modifying it.

If you'd prefer not to bother with that (although I assure you there is really not much to come to grips with) and just write it all over again from scratch then that is of course your prerogative. I find it disappointing that your reaction is to refer to it as _"ridiculously overengineered"_ - though irrespective of this , I will move forward with formally documenting the new code structure and you can choose to review it (or not)

As for breaking compatibility with the existing code and on the topic of actually merging it, I again refer to my original statement:

> Since I haven't yet implemented migrations or the remote interface, it's definitely not ready for primetime

Once the above has been resolved, I believe that the GUI-based layout profiling feature certainly has significant value over the existing Factorissimo, I will make it into its own fork if you'd rather go your own way.
